docs/exceptions.md

```diff
@@ -155,6 +181,11 @@ Do:
 * use `MonadError` or `MonadThrow` (methods of these classes). Note
   that if you define `f :: MonadError m => m ()`, it won't be pure
 
+*Comment:* I disagree with the last "Do not". Why is this ruled out? Sometimes
+giving `Either Text` is legit. Giving `Maybe` is ok, so it must also ok to
+give `Either ()`, so why not `Either Text`, which contains even more
+information?
+
 Consider parsing: it is pure, but we cannot make assumptions about the input. In
```

Comment:
We *do* make assumptions about the input, namely its type. We do *not* assume that the input is recognised by the parser. But of course we don't! We'd have to define the parser in order to do that.

Reply:
> We do make assumptions about the input, namely its type.

We don't make any assumptions about the input, the input is whatever passed to our programs, and the type we choose to represent it (`String`, `Text`, `ByteString`, `Vector Char`, or whatever) isn't an assumption, it's an engineering decision.

Reply:
No, it's an assumption too. If we choose `Text`, we know we don't have to handle a `True` as input.

Reply:
`True` as a Haskell isn't something a user can enter, parsing input normally comes from a file or from REPL, so the input is necessarily a sequence of characters. We could also run a lexer as a preprocessing pass, then the input would be a sequence of tokens. Or we could be dealing with a binary format, in which case the input would be a sequence of bytes.

In any case, we're not making any assumptions about the input to the program.

Reply:
> In any case, we're not making any assumptions about the input to the program.

But you just admitted that we do make assumptions.

> the input is necessarily a sequence of characters

Reply:
I propose the following clarification:

```
- but we cannot make assumptions about the input
+ but we cannot make assumptions about the input string
```

Reply:
Sure, but why make this observation at all? We can't assume that the input string is recognised by our parser, but that's why we're parsing it: we want to determine whether it's recognised by our parser. Would you define

```Haskell
-- We can't assume the input is 42, so we throw an exception if it isn't.
is42 :: Int -> ExceptT Not42 ()`
is42 42 = pure ()
is42 _ = throwError Not42
```

Can we agree that this is a silly thing to write?

Reply:
> We can't assume that the input string is recognised by our parser, but that's why we're parsing it: we want to determine whether it's recognised by our parser.

No, that's not why we're parsing it. We're parsing it because we want to extract data from it. We can assume it is recognized by the parser: if our nodes communicate via JSON, the bytes sent over the network must adhere to the JSON grammar, otherwise it's an error (or an attack, but clearly not normal operation).

Reply:
> We're parsing it because we want to extract data from it. 

Right, in order to do so, it must be well-formed, recognised by our parser, so that we can map it to the structure we want.

> We can assume it is recognized by the parser: if our nodes communicate via JSON, the bytes sent over the network must adhere to the JSON grammar, otherwise it's an error (or an attack, but clearly not normal operation).

Ok let's keep with this example of a particular application of a parser. What is the advantage of throwing an exception on unrecognised input (impure), versus returning a value (pure)?

Reply:
> Ok let's keep with this example of a particular application of a parser. What is the advantage of throwing an exception on unrecognised input (impure), versus returning a value (pure)?

Disadvantages of exceptions as values are described in "A Semantics for Imprecise Exceptions", section 2.2, "Inadequacies of exceptions as values". Screenshot attached

![image](https://user-images.githubusercontent.com/4061728/36114495-38ae854e-1041-11e8-9d36-ef4ab285114d.png)

Reply:
Of course, with parsers we often need to backtrack on failure, and we don't make the assumption that the input adheres to the grammar, and parsing is a pure operation, so the best option is `ExceptT`. This is why the document refers to parsing as an example where we want to represent errors as values, not throw them.